packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol

```diff
@@ -0,0 +1,135 @@
+
+    /// @notice Nonce for the next message to be sent, without the message version applied. Use the
+    ///         messageNonce getter which will insert the message version into the nonce to give you
+    ///         the actual nonce to be used for the message.
+    uint240 internal msgNonce;
+
+    /// @notice Emitted whenever a message is sent to the other chain.
+    /// @param destination Chain ID of the destination chain.
+    /// @param target      Target contract or wallet address.
+    /// @param message     Message to trigger the target address with.
+    /// @param data        Data to be sent with the message.
+    event SentMessage(uint256 destination, address target, bytes message, bytes data) anonymous;
+
+    /// @notice Retrieves the next message nonce. Message version will be added to the upper two
+    ///         bytes of the message nonce. Message version allows us to treat messages as having
+    ///         different structures.
+    /// @return Nonce of the next message to be sent, with added message version.
+    function messageNonce() public view returns (uint256) {
+        return Encoding.encodeVersionedNonce(msgNonce, MESSAGE_VERSION);
+    }
+
+    /// @notice Sends a message to some target address on a destination chain. Note that if the call
+    ///         always reverts, then the message will be unrelayable, and any ETH sent will be
+    ///         permanently locked. The same will occur if the target on the other chain is
+    ///         considered unsafe (see the _isUnsafeTarget() function).
+    /// @param _destination Chain ID of the destination chain.
+    /// @param _target      Target contract or wallet address.
+    /// @param _message     Message to trigger the target address with.
+    function sendMessage(uint256 _destination, address _target, bytes calldata _message) external payable {
+        require(_destination != block.chainid, "L2ToL2CrossDomainMessenger: cannot send message to self");
+
+        bytes memory data = abi.encodeCall(
+            L2ToL2CrossDomainMessenger.relayMessage,
+            (_destination, messageNonce(), msg.sender, _target, msg.value, _message)
+        );
+        emit SentMessage(_destination, _target, _message, data);
+        msgNonce++;
+    }
+
+    /// @notice Relays a message that was sent by the other CrossDomainMessenger contract. Can only
+    ///         be executed via cross-chain call from the other messenger OR if the message was
+    ///         already received once and is currently being replayed.
+    /// @param _destination Chain ID of the destination chain.
+    /// @param _nonce       Nonce of the message being relayed.
+    /// @param _sender      Address of the user who sent the message.
+    /// @param _target      Address that the message is targeted at.
+    /// @param _value       ETH value to send with the message.
+    /// @param _message     Message to send to the target.
+    function relayMessage(
+        uint256 _destination,
```

Comment:
This needs to be updated to include https://github.com/ethereum-optimism/specs/pull/81